crypto/ube/ube.c

```diff
@@ -0,0 +1,238 @@
+// needed because currently the mutation is only ever assigning 1 to the
+// variable.
+static void ube_failed(void) {
+  CRYPTO_once(&ube_detection_unavailable_once, set_ube_detection_unavailable_once);
+}
+
+// ube_state_initialize initalizes the global state |ube_state|.
+static void ube_state_initialize(void) {
+
+  ube_state.generation_number = 0;
+  ube_state.cached_fork_gn = get_fork_generation_number_mocked();
+  int ret_snapsafe_gn =
+    get_snapsafe_generation_number_mocked(&(ube_state.cached_snapsafe_gn));
+
+  if (ube_state.cached_fork_gn == 0 ||
+      ret_snapsafe_gn == 0) {
+    ube_failed();
+  }
+}
+
+// ube_update_state updates the global state |ube_state| with the generation
+// numbers loaded into |current_detection_gn| and increments the UBE generation
+// number.
+static void ube_update_state(struct detection_gn *current_detection_gn) {
+
+  ube_state.generation_number += 1;
+
+  // Make sure we cache all new generation numbers. Otherwise, we might detect
+  // a fork UBE but, in fact, both a fork and snapsafe UBE occurred. Then next
+  // time we enter, a redundant reseed will be emitted.
+  ube_state.cached_fork_gn = current_detection_gn->current_fork_gn;
+  ube_state.cached_snapsafe_gn = current_detection_gn->current_snapsafe_gn;
+}
+
+// ube_get_detection_generation_numbers loads the current detection generation
+// numbers into |current_detection_gn|.
+//
+// Returns 1 on success and 0 otherwise. The 0 return value means that a
+// detection method we expected to be available, is in fact not.
+static int ube_get_detection_generation_numbers(
+  struct detection_gn *current_detection_gn) {
+
+  // An attempt to prevent a situation where new detection methods are added but
+  // we forget to load them.
+  OPENSSL_STATIC_ASSERT(NUMBER_OF_DETECTION_GENERATION_NUMBERS == 2,
+    not_handling_the_exact_number_of_detection_generation_numbers);
+
+  current_detection_gn->current_fork_gn = get_fork_generation_number_mocked();
+  int ret_snapsafe_gn = get_snapsafe_generation_number_mocked(
+                          &(current_detection_gn->current_snapsafe_gn));
+
+  if (current_detection_gn->current_fork_gn == 0 ||
+      ret_snapsafe_gn == 0) {
+    return 0;
+  }
+
+  return 1;
+}
+
+// ube_is_detected computes whether a UBE has been detected or not by comparing
+// the cached detection generation numbers with the current detection generation
+// numbers. The current generation numbers must be loaded into
+// |current_detection_gn| before calling this function.
+//
+// Returns 1 if UBE has been detected and 0 if no UBE has been detected.
+static int ube_is_detected(struct detection_gn *current_detection_gn) {
+
+  if (ube_state.cached_fork_gn != current_detection_gn->current_fork_gn ||
+      ube_state.cached_snapsafe_gn != current_detection_gn->current_snapsafe_gn) {
+    return 1;
+  }
+  return 0;
+}
+
+// stays invariant after a failure state has been reached.
```

Comment:
This feels like an incomplete thoughts, what stays invariant?

Reply:
forgot to remove...